vowpalwabbit/learner.h

```diff
@@ -128,6 +135,174 @@ inline void decrement_offset(multi_ex& ec_seq, const size_t increment, const siz
+    TYPE saved = std::move(ec.l.FIELD_NAME);                                                                         \
+    learn_fd.learn_with_label_f(learn_fd.data, *learn_fd.base, static_cast<void*>(&ec), static_cast<void*>(&saved)); \
+    ec.l.FIELD_NAME = std::move(saved);                                                                              \
+  }                                                                                                                  \
+  break;
+
+#define TYPED_MULTI_EXAMPLE_LEARN_CASE(TYPE, FIELD_NAME)                                                             \
+  case label_type_t::FIELD_NAME:                                                                                     \
+  {                                                                                                                  \
+    std::vector<TYPE> saved;                                                                                         \
+    saved.reserve(ec.size());                                                                                        \
+    for (const auto& example : ec)                                                                                   \
+    {                                                                                                                \
+      saved.push_back(std::move(example->l.FIELD_NAME));                                                             \
+    }                                                                                                                \
+    learn_fd.learn_with_label_f(learn_fd.data, *learn_fd.base, static_cast<void*>(&ec), static_cast<void*>(&saved)); \
+    for (size_t i = 0; i < ec.size(); i++)                                                                           \
+    {                                                                                                                \
+      const auto& example = ec[i];                                                                                   \
+      example->l.FIELD_NAME = std::move(saved[i]);                                                                   \
+    }                                                                                                                \
+  }                                                                                                                  \
+  break;
+
+#define TYPED_EXAMPLE_PREDICT_CASE(TYPE, FIELD_NAME)                                                                   \
+  case label_type_t::FIELD_NAME:                                                                                       \
+  {                                                                                                                    \
+    TYPE saved = std::move(ec.l.FIELD_NAME);                                                                           \
```

Comment:
>move [](start = 22, length = 4)

Why is this needed during a predict?

Reply:
In the current state of the codebase label information is required in predict. This will be tackled in this feature branch before we are done.